File: leetcode/go/2696.minimum-string-length-after-removing-substrings.go
func minLength(s string) int {
    runes := []rune(s)

    stack := make([]rune, 0)
    l := -1

    for _, r := range runes {
        if l > -1 && ((r == 'D' && stack[l] == 'C') || (r == 'B' && stack[l] == 'A')) {
            stack = stack[0:l]
            l--
        } else {
            stack = append(stack, r)
            l++
        }
    }

    return len(stack)
}

File: leetcode/cpp/2696.minimum-string-length-after-removing-substrings.cpp
class Solution {
public:
    int minLength(string s) {
        vector<char> stack;
        for (const char c: s) {
            if (!stack.empty() && ((stack.back() == 'C' && c == 'D') || (stack.back() == 'A' && c == 'B'))) {
                stack.pop_back();
            } else {
                stack.push_back(c);
            }
        }
        return stack.size();
    }
};

/* */

class Solution {
public:
    int minLength(string s) {
        string stack = "";
        for (const char c: s) {
            if (!stack.empty() && ((stack.back() == 'C' && c == 'D') || (stack.back() == 'A' && c == 'B'))) {
                stack.pop_back();
            } else {
                stack.push_back(c);
            }
        }
        return stack.length();
    }
};

